From aa054e76ced01b62ee98b098b93b107f3fc854fb Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Tue, 14 Oct 2025 09:52:01 +0200
Subject: [PATCH 412/553] QUndoStack: use idiomatic relational operators

Don't only define the inequality operator. Not even C++20 will
synthesize equality from it. Instead, for C++20, default equality and,
for non-C++20-builds, define inequality in terms of non-equality.

Also =delete qHash(). Equality-comparable types ought to be hashable,
but the code currently doesn't need this feature, so document this by
deleting the qHash overload explicitly. This also prevents 3rd-parties
from defining their own qHash(ActionState), an issue of which we have
had several instances in Qt already, and which can lead to ODR
violations.

Amends 31b0dadb0f371fc94652782c28024f135a0b6f4b.

Pick-to: 6.8
Task-number: QTBUG-138567
Change-Id: Ib766d44697cb763147eafceb5ffd947206bda979
Reviewed-by: Ivan Solovev <ivan.solovev@qt.io>
(cherry picked from commit 9b4c15ef5ffa94829a0c690eba37472c6833600c)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/gui/util/qundostack_p.h | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/src/gui/util/qundostack_p.h b/src/gui/util/qundostack_p.h
index fea201ce62d..6bdcf5fb20b 100644
--- a/src/gui/util/qundostack_p.h
+++ b/src/gui/util/qundostack_p.h
@@ -59,10 +59,17 @@ public:
         bool enabled = false;
         QString text;
 
-        bool operator!=(const ActionState &other) const noexcept
-        {
-            return enabled != other.enabled || text != other.text;
-        }
+        friend bool operator==(const ActionState &lhs, const ActionState &rhs) noexcept
+#ifdef __cpp_impl_three_way_comparison
+            = default;
+#else
+        { return lhs.enabled == rhs.enabled && lhs.text == rhs.text; }
+        friend bool operator!=(const ActionState &lhs, const ActionState &rhs) noexcept
+        { return !(lhs == rhs); }
+#endif
+        // some compiler's reject seed = 0) = delete, overload instead:
+        friend void qHash(const ActionState &key, size_t seed) = delete;
+        friend void qHash(const ActionState &key) = delete;
     };
 
     QList<QUndoCommand*> command_list;
-- 
2.51.2

